xds/internal/server/rds_handler.go

```diff
@@ -0,0 +1,127 @@
+
+	updateChannel chan rdsHandlerUpdate
+}
+
+// newRDSHandler is expected to called once on instantiation of a wrapped
+// listener. On any LDS updates the wrapped listener receives, the listener
+// should update the handler with the route names (which specify dynamic RDS)
+// using the function below.
+func newRDSHandler(parent *listenerWrapper) *rdsHandler {
+	return &rdsHandler{
+		parent:        parent,
+		updateChannel: make(chan rdsHandlerUpdate, 1),
+		rdsUpdates:    make(map[string]xdsclient.RouteConfigUpdate),
+		rdsCancels:    make(map[string]func()),
+	}
+}
+
+// updateRouteNamesToWatch handles a list of route names to watch for a given
+// server side listener (if a filter chain specifies dynamic RDS configuration).
+// This function handles all the logic with respect to any routes that may have
+// been added or deleted as compared to what was previously present.
+func (rh *rdsHandler) updateRouteNamesToWatch(routeNamesToWatch map[string]bool) {
+	rh.rdsMutex.Lock()
+	defer rh.rdsMutex.Unlock()
+	// Add and start watches for any routes for any new routes in routeNamesToWatch.
+	for routeName := range routeNamesToWatch {
+		if _, ok := rh.rdsCancels[routeName]; !ok {
+			rh.rdsCancels[routeName] = rh.parent.xdsC.WatchRouteConfig(routeName, rh.handleRouteUpdate)
+		}
+	}
+
+	// Delete and cancel watches for any routes from persisted routeNamesToWatch
+	// that are no longer present.
+	for routeName := range rh.rdsCancels {
+		if _, stillRDS := routeNamesToWatch[routeName]; !stillRDS {
+			rh.rdsCancels[routeName]()
+			delete(rh.rdsCancels, routeName)
+			delete(rh.rdsUpdates, routeName)
+		}
+	}
+}
+
+// handleRouteUpdate persists the route config for a given route name, and also
+// sends an update to the Listener Wrapper on an error received or if the rds
+// handler has a full collection of updates.
+func (rh *rdsHandler) handleRouteUpdate(update xdsclient.RouteConfigUpdate, err error) {
+	rh.rdsMutex.Lock()
+	defer rh.rdsMutex.Unlock()
+	// Note: this doesn't need a check for if name in RouteConfigUpdate is wrong
+	// (i.e. not started a watch for). This will be validated in xdsclient.
+	if err != nil {
+		// For a rdsHandler update, the only update wrapped listener cares about
+		// is most recent one, so opportunistically drain the update before
+		// sending the new update.
+		select {
+		case <-rh.updateChannel:
+		default:
+		}
+		rh.updateChannel <- rdsHandlerUpdate{err: err}
+		return
+	}
+	rh.rdsUpdates[update.RouteName] = update
+
+	// If the full list (determined by length) of rdsUpdates have successfully updated,
+	// the listener is ready to be updated.
+	if len(rh.rdsUpdates) == len(rh.rdsCancels) {
+		// For a rdsHandler update, the only update lis wrapper cares about is most recent one,
+		// so opportunistically drain the update before sending the new update.
+		select {
+		case <-rh.updateChannel:
+		default:
+		}
+		rh.updateChannel <- rdsHandlerUpdate{rdsUpdates: rh.rdsUpdates}
+	}
```

Comment:
Would the following case be covered?

- `updateRouteNamesToWatch(A, B, C)`
- `WatchRouteConfig()` is called for A, B and C
- We get a response for `A` and `B`
- Still the condition `len(rh.rdsUpdates) == len(rh.rdsCancels)` is not true, so no update is propagated to the `listenerWrapper`
- Then `updateRouteNamesToWatch(A, B)` happens
- map entries for `C` are deleted
- Now, the above condition is true, but the update will not be propagated up. Am I right?

I would prefer having a single map:
```golang
type updateWithCancel struct {
  update *xdsclient.RouteConfigUpdate
  cancel func()
}

type rdsHandler struct {
  ...
  updates map[string]updateWithCancel
  ...
}

func (rh *rdsHandler) readyToUpdate() bool { // or a better name
  // Traverses the map and figures out if we are ready to send an update upstairs
}
```

Call `readyToUpdate` from both `updateRouteNamesToWatch` and `handleRouteUpdate` and send update upstairs if required.

Reply:
From your perspective, why would a map[string]{A, B} be cleaner than map[string]{A} and a map[string]{B}? I agree with your correctness issue. However, traversing a single map would add a 0(n) search for every update/update route name to watch vs len() and len(), which is constant.